### DISCUSSION TITLE
Consume from routingKey

### DISCUSSION THREAD

**Discussion by guicuton** [php-amqplib/php-amqplib#1017] [Help wanted]
Hey guys,

It's possible to consume messages from a given routingKey of some queue?
My queue has messages with the following routingKeys: 
- callback
- logs

I've tried to bind queue with the routingKey and then `basic_get` but the messages returned isn't "filtered"

**Example:**
```php
$channel->queue_bind('my_queue', 'amq.direct', 'callback');
$message = $channel->basic_get('my_queue');

var_dump($message); //Message with routing key logs are returned despite the bind on callback
```
_Ps: Please, do not comment about the "avoid using basic_get". This is not the point._

There  is a way to do it using the methods or i'll need to consume first and them filter the messages by their routingKey?

**Comment by lukebakken** [comment]
This is not supported by RabbitMQ.

You should use topic routing to route messages to designated queues. In your case, one for `callback` and one for `logs`.

https://www.rabbitmq.com/tutorials/tutorial-five-php.html

**Comment by guicuton** [comment]
Hi @lukebakken thank you for your reply!

This is what i'm trying to avoid to do.
 I'm working under a requirement that these informations should exist only in one queue.

Since this is not supported for RabbitMQ the best way is to consume the queue and filter the message by this data, right?
```php
$channel->queue_bind('my_queue', 'amq.direct');
$channel->basic_consume('my_queue', '', false, false, false, false, function($message) {
  /** 
    * here do the ifs according the routingKey and 
    * populate some array of each one to use it as am i needed, right?
    * /
});

**Reply by lukebakken** [reply]
Yes, but you will have to "nack" the messages you don't want, or else they will be in the "unacked" state until your consuming channel closes.

Also, note that the messages may not be re-queued in exactly the same order when nack-ed. So if you're using one queue to preserve message order, you may be out of luck anyway and should just use multiple queues.

**Reply by guicuton** [reply]
Thank you again @lukebakken 

The order isn't important.

Thank you very much mate!
